cmd/ctl/pkg/inspect/secret/secret.go

```diff
@@ -0,0 +1,382 @@
+	}
+
+	certData := secret.Data[corev1.TLSCertKey]
+	certs, err := splitPEMs(certData)
+	if err != nil {
+		return err
+	}
+	if len(certs) < 1 {
+		return errors.New("no PEM data found in secret")
+	}
+
+	intermediates := [][]byte(nil)
+	if len(certs) > 1 {
+		intermediates = certs[1:]
+	}
+
+	// we only want to inspect the leaf certificate
+	x509Cert, err := pki.DecodeX509CertificateBytes(certs[0])
+	if err != nil {
+		return fmt.Errorf("error when parsing 'tls.crt': %w", err)
+	}
+
+	out := []string{
+		describeValidFor(x509Cert),
+		describeValidityPeriod(x509Cert),
+		describeIssuedBy(x509Cert),
+		describeIssuedFor(x509Cert),
+		describeCertificate(x509Cert),
+		describeDebugging(x509Cert, intermediates, secret.Data[cmmeta.TLSCAKey]),
+	}
+
+	fmt.Println(strings.Join(out, "\n\n"))
+
+	return nil
+}
+
+func describeValidFor(cert *x509.Certificate) string {
+	var b bytes.Buffer
+	template.Must(template.New("validForTemplate").Parse(validForTemplate)).Execute(&b, struct {
+		DNSNames       string
+		URIs           string
+		IPAddresses    string
+		EmailAddresses string
+		KeyUsage       string
+	}{
+		DNSNames:       printSlice(cert.DNSNames),
+		URIs:           printSlice(pki.URLsToString(cert.URIs)),
+		IPAddresses:    printSlice(pki.IPAddressesToString(cert.IPAddresses)),
+		EmailAddresses: printSlice(cert.EmailAddresses),
+		KeyUsage:       printKeyUsage(pki.BuildCertManagerKeyUsages(cert.KeyUsage, cert.ExtKeyUsage)),
+	})
+
+	return b.String()
+}
+
+func describeValidityPeriod(cert *x509.Certificate) string {
+	var b bytes.Buffer
+	template.Must(template.New("validityPeriodTemplate").Parse(validityPeriodTemplate)).Execute(&b, struct {
+		NotBefore string
+		NotAfter  string
+	}{
+		NotBefore: cert.NotBefore.Format(time.RFC1123),
+		NotAfter:  cert.NotAfter.Format(time.RFC1123),
+	})
+
+	return b.String()
+}
+
+func describeIssuedBy(cert *x509.Certificate) string {
+	var b bytes.Buffer
+	template.Must(template.New("issuedByTemplate").Parse(issuedByTemplate)).Execute(&b, struct {
+		CommonName         string
+		Organization       string
+		OrganizationalUnit string
+		Country            string
+	}{
+		CommonName:         printOrNone(cert.Issuer.CommonName),
+		Organization:       printSliceOrOne(cert.Issuer.Organization),
+		OrganizationalUnit: printSliceOrOne(cert.Issuer.Organization),
+		Country:            printSliceOrOne(cert.Issuer.Country),
+	})
+
+	return b.String()
+}
+
+func describeIssuedFor(cert *x509.Certificate) string {
+	var b bytes.Buffer
+	template.Must(template.New("issuedForTemplate").Parse(issuedForTemplate)).Execute(&b, struct {
+		CommonName         string
+		Organization       string
+		OrganizationalUnit string
+		Country            string
+	}{
+		CommonName:         printOrNone(cert.Subject.CommonName),
+		Organization:       printSliceOrOne(cert.Subject.Organization),
+		OrganizationalUnit: printSliceOrOne(cert.Subject.Organization),
+		Country:            printSliceOrOne(cert.Subject.Country),
+	})
```

Comment:
I think these `template.Must(template.New("xxx").Parse(xxx))` should be moved to vars at the top.

Reply:
+1